File: 312. Burst Balloons.py
# 312. Burst Balloons

class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]
        n = len(nums)
        dp = [[0] * n for _ in range(n)]

        for left in range(n-2, 0, -1): # 터뜨릴 수 있는 가장  왼쪽거
            for right in range(left, n-1): # 터트릴 가장 오른쪽거
                for i in range(left, right + 1): # 그거를 왼쪽부터 터뜨려보자
                    last_pop =  nums[left-1] * nums[i] * nums[right+1]
                    remain = dp[left][i-1] + dp[i+1][right] # 방금 계산 한거 + 아까 계산한 아랫줄(이 풍선 없는 조합)
                    dp[left][right] = max(dp[left][right], last_pop + remain)
            
        return dp[1][n-2]

File: 312. Burst Balloons.cpp
// 312. Burst Balloons
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
    public:
        int maxCoins(vector<int>& nums) {
            vector<int> balloons(nums.size()+2, 1);
            int n = balloons.size();
    
            for (int i = 1; i < n-1; i++) {
                balloons[i] = nums[i-1];
            }
    
            vector<vector<int>> dp(n, vector<int>(n, 0));
    
            for (int left = n-2; left > 0; left--) {
                for (int right = left; right < n-1; right++) {
                    for (int i = left; i < right + 1; i++) {
                        int lastPop = balloons[left-1] * balloons[i] * balloons[right+1];
                        int remains = dp[left][i-1] + dp[i+1][right];
                        dp[left][right] = max(dp[left][right], lastPop + remains);
                    }
                }
            }
    
            return dp[1][n-2];
    
        }
    };